File: android/src/main/java/com/voicekit/VoiceError.kt
package com.voicekit

sealed class VoiceError(
  val code: String,
  message: String
) : Exception(message) {
  object SpeechRecognizerNotAvailable : VoiceError(
    "ERR_SPEECH_RECOGNIZER_NOT_AVAILABLE",
    "Speech recognition is not available on this device"
  )
  object RecordingStartFailed : VoiceError(
    "ERR_RECORDING_START_FAILED",
    "Failed to start audio recording"
  )
  object RecognitionFailed : VoiceError(
    "ERR_RECOGNITION_FAILED",
    "Failed to recognize speech"
  )
  object PermissionDenied : VoiceError(
    "ERR_PERMISSION_DENIED",
    "Speech recognition permission was denied"
  )
  object PermissionRestricted : VoiceError(
    "ERR_PERMISSION_RESTRICTED",
    "Speech recognition is restricted on this device"
  )
  object PermissionNotDetermined : VoiceError(
    "ERR_PERMISSION_NOT_DETERMINED",
    "Speech recognition permission was not yet determined"
  )
  object InvalidState : VoiceError(
    "ERR_INVALID_STATE",
    "Invalid state, cannot perform action"
  )
  class Unknown(message: String = "An unknown error occurred") : VoiceError(
    "ERR_UNKNOWN",
    message
  )
}

File: ios/Utils/VoiceError.swift
//
//  VoiceError.swift
//  VoiceKit
//
//  Created by Maximilian Krause on 20.11.24.
//  Copyright © 2024 Kuatsu App Agency. All rights reserved.
//

import Foundation

enum VoiceError: Error {
  case speechRecognizerNotAvailable
  case recordingStartFailed
  case recognitionFailed
  case permissionDenied
  case permissionRestricted
  case permissionNotDetermined
  case invalidState
  case unknown(message: String = "An unknown error occurred")

  var code: String {
    switch self {
    case .speechRecognizerNotAvailable: "ERR_SPEECH_RECOGNIZER_NOT_AVAILABLE"
    case .recordingStartFailed: "ERR_RECORDING_START_FAILED"
    case .recognitionFailed: "ERR_RECOGNITION_FAILED"
    case .permissionDenied: "ERR_PERMISSION_DENIED"
    case .permissionRestricted: "ERR_PERMISSION_RESTRICTED"
    case .permissionNotDetermined: "ERR_PERMISSION_NOT_DETERMINED"
    case .invalidState: "ERR_INVALID_STATE"
    case .unknown: "ERR_UNKNOWN"
    }
  }

  var message: String {
    switch self {
    case .speechRecognizerNotAvailable:
      "Speech recognition is not available on this device"
    case .recordingStartFailed:
      "Failed to start audio recording"
    case .recognitionFailed:
      "Failed to recognize speech"
    case .permissionDenied:
      "Speech recognition permission was denied"
    case .permissionRestricted:
      "Speech recognition is restricted on this device"
    case .permissionNotDetermined:
      "Speech recognition permission was not yet determined"
    case .invalidState:
      "Invalid state, cannot perform action"
    case let .unknown(message):
      message
    }
  }
}
